pmd-java/src/main/resources/category/java/errorprone.xml

```diff
@@ -1116,6 +1116,58 @@ public void foo() {
         </example>
     </rule>
 
+    <rule name="DetachedTestCase"
+          language="java"
+          since="6.13.0"
+          message="Probable detached JUnit test case."
+          class="net.sourceforge.pmd.lang.rule.XPathRule"
+          typeResolution="true"
+          externalInfoUrl="${pmd.website.baseurl}/pmd_rules_java_errorprone.html#detachedtestcase">
+        <description>
+The method appears to be a test case since it has public or default visibility,
+non-static access, no arguments, no return value, has no annotations, but is a
+member of a class that has one or more JUnit test cases. If it is a utility
+method, it should likely have private visibility. If it is an ignored test, it
+should be annotated with @Test and @Ignore.
+        </description>
+        <priority>3</priority>
+        <properties>
+            <property name="xpath">
+                <value><![CDATA[
+//ClassOrInterfaceBodyDeclaration
+[../ClassOrInterfaceBodyDeclaration/Annotation/*/Name
+        [pmd-java:typeIs('org.junit.Test')
+         or pmd-java:typeIs('org.junit.jupiter.api.Test')
+         or pmd-java:typeIs('org.junit.jupiter.api.RepeatedTest')
+         or pmd-java:typeIs('org.junit.jupiter.api.TestFactory')
+         or pmd-java:typeIs('org.junit.jupiter.api.TestTemplate')
+         or pmd-java:typeIs('org.junit.jupiter.params.ParameterizedTest')]
+]
+[not(Annotation)]
+[MethodDeclaration[(@Public = 'true' or @PackagePrivate = 'true') and @Static = 'false' and
```

Comment:
Just a suggestion! should we also make sure that this method's body has at least one Assertion or an expected exception so that any unforeseen false positive is avoided?

Reply:
I don't think, that's necessary: True, the rule will flag any public method in the test class, which is not annotated as test. But then, you need to think about: why is this method public in the first place? As stated in the rule description, that method should probably be private or static...

I know, there is at least one false positive in spring framework detected:
https://github.com/spring-projects/spring-framework/blob/v5.0.6.RELEASE/spring-aop/src/test/java/org/springframework/aop/aspectj/AspectJAdviceParameterNameDiscovererTests.java#L303

But this looks like test data (that could be in an extra separate class) is mixed with the test itself, making the test class unnecessarily big.

Other false positives:
* https://github.com/spring-projects/spring-framework/blob/v5.0.6.RELEASE/spring-expression/src/test/java/org/springframework/expression/spel/VariableAndFunctionTests.java#L93
* https://github.com/spring-projects/spring-framework/blob/v5.0.6.RELEASE/spring-test/src/test/java/org/springframework/test/web/servlet/result/PrintingResultHandlerTests.java#L379

So: 3 false positives, 11 real detached test cases. The 3 false positives are all not externalized test data mixed with the test itself.

Reply:
Good point! Agree with your view to have all the unannotated public test data methods in a separate class.